pkg/auth/providers/azure/azure_provider.go

```diff
@@ -211,154 +217,154 @@ func (ap *azureProvider) TransformToAuthProvider(
-
-	azureClient, err := newClientCode(azureCredential.Code, config)
-	if err != nil {
-		return v3.Principal{}, nil, "", err
+	if authProviderOld(config) {
+		return ap.loginUserAzureADGraph(azureCredential, config, test)
 	}
+	return ap.loginUserMSGraph(azureCredential, config, test)
+}
 
-	logrus.Debug("[AZURE_PROVIDER] Completed token swap with AzureAD")
-
-	oid, err := parseJWTforField(azureClient.accessToken(), "oid")
+func (ap *azureProvider) getUser(client clients.AzureClient, principalID string, token v3.Token) (v3.Principal, error) {
+	u, err := client.GetUser(principalID)
 	if err != nil {
-		return v3.Principal{}, nil, "", err
+		return v3.Principal{}, httperror.NewAPIError(httperror.NotFound, err.Error())
 	}
 
-	logrus.Debug("[AZURE_PROVIDER] Started getting user info from AzureAD")
-
-	user, err := azureClient.userClient.Get(context.Background(), oid)
+	p, err := userToPrincipal(u)
 	if err != nil {
-		return v3.Principal{}, nil, "", err
+		return v3.Principal{}, err
 	}
+	p.Me = samePrincipal(&token.UserPrincipal, p)
 
-	logrus.Debug("[AZURE_PROVIDER] Completed getting user info from AzureAD")
-
-	userPrincipal := ap.userToPrincipal(user)
-	userPrincipal.Me = true
-
-	var mem bool
-	params := graphrbac.UserGetMemberGroupsParameters{
-		SecurityEnabledOnly: &mem,
-	}
+	return *p, nil
+}
 
-	userGroups, err := azureClient.userClient.GetMemberGroups(context.Background(), *user.ObjectID, params)
+func (ap *azureProvider) getGroup(client clients.AzureClient, id string, token v3.Token) (v3.Principal, error) {
+	group, err := client.GetGroup(id)
 	if err != nil {
-		return v3.Principal{}, nil, "", err
+		return v3.Principal{}, httperror.NewAPIError(httperror.NotFound, err.Error())
 	}
 
-	groupPrincipals, err := ap.userGroupsToPrincipals(azureClient, userGroups)
+	p, err := groupToPrincipal(group)
 	if err != nil {
-		return v3.Principal{}, nil, "", err
+		return v3.Principal{}, err
 	}
+	p.MemberOf = ap.tokenMGR.IsMemberOf(token, *p)
 
-	testAllowedPrincipals := config.AllowedPrincipalIDs
-	if test && config.AccessMode == "restricted" {
-		testAllowedPrincipals = append(testAllowedPrincipals, userPrincipal.Name)
-	}
+	return *p, nil
+}
 
-	allowed, err := ap.userMGR.CheckAccess(config.AccessMode, testAllowedPrincipals, userPrincipal.Name, groupPrincipals)
+func (ap *azureProvider) searchUsers(client clients.AzureClient, name string, token v3.Token) ([]v3.Principal, error) {
+	filter := fmt.Sprintf("startswith(userPrincipalName,'%[1]s') or startswith(displayName,'%[1]s') or startswith(givenName,'%[1]s') or startswith(surname,'%[1]s')", name)
+	users, err := client.ListUsers(filter)
 	if err != nil {
-		return v3.Principal{}, nil, "", err
-	}
-	if !allowed {
-		return v3.Principal{}, nil, "", httperror.NewAPIError(httperror.Unauthorized, "unauthorized")
+		return nil, err
 	}
 
-	providerToken, err := azureClient.marshalTokenJSON()
-	if err != nil {
-		return v3.Principal{}, nil, "", err
+	switch u := users.(type) {
```

Comment:
I would recommend adjusting the relevant functions of the clients to return a list of principals rather than an interface{} which forces the provider to do the conversion for them.